src/private_broadcast.h

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2023-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://opensource.org/license/mit/.
+
+#ifndef BITCOIN_PRIVATE_BROADCAST_H
+#define BITCOIN_PRIVATE_BROADCAST_H
+
+#include <primitives/transaction.h>
+#include <sync.h>
+#include <threadsafety.h>
+#include <util/hasher.h>
+#include <util/time.h>
+
+#include <map>
+#include <unordered_map>
+
+/**
+ * Store a list of transactions to be broadcast privately. Supports the following operations:
+ * - Add a new transaction
+ */
+class PrivateBroadcast
+{
+public:
+    /**
+     * Add a transaction to the storage.
+     * @param[in] tx The transaction to add.
+     * @retval true The transaction was added.
+     * @retval false The transaction was already present.
+     */
+    bool Add(const CTransactionRef& tx) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
+
+private:
+    struct Priority {
+        size_t num_broadcasted{0};
+        NodeClock::time_point last_broadcasted{};
+
+        bool operator<(const Priority& other) const;
+    };
+
+    struct TxWithPriority {
+        CTransactionRef tx;
+        Priority priority;
+    };
+
+    using ByTxid = std::unordered_map<Txid, TxWithPriority, SaltedTxidHasher>;
```

Comment:
Because if stored by wtxid then the same could happen, but the attacker malleates the transaction to a valid witness data, so it has same txid, different wtxid, and is valid. When we receive that transaction we would not recognize it because we look at the wtxid, so we would keep broadcasting the original one, until the code in `PeerManagerImpl::ReattemptPrivateBroadcast()` gives up on it because it is not acceptable in the mempool.

Do you think we should check both txid and wtxid when we receive a transaction from the network, or try to validate the transaction before removing it via `m_tx_for_private_broadcast.Remove(ptx)`?

Reply:
I'd say it might be a good idea moving the removal until after validation - that should address both scenarios.
We are going to validate it anyway, so I don't really see any downsides to that.

Reply:
Here is the change:

<details>
<summary>[patch] only remove if accepted in mempool</summary>

```diff
--- i/src/net_processing.cpp
+++ w/src/net_processing.cpp
@@ -4803,27 +4803,12 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         const uint256& txid = ptx->GetHash();
         const uint256& wtxid = ptx->GetWitnessHash();
 
         const uint256& hash = peer->m_wtxid_relay ? wtxid : txid;
         AddKnownTx(*peer, hash);
 
-        if (auto num_broadcasted = m_tx_for_private_broadcast.Remove(ptx)) {
-            LogPrintLevel(BCLog::PRIVATE_BROADCAST,
-                          BCLog::Level::Info,
-                          "Received our privately broadcast transaction (txid=%s) from the "
-                          "network from peer=%d%s; stopping private broadcast attempts\n",
-                          txid.ToString(),
-                          pfrom.GetId(),
-                          fLogIPs ? strprintf(", peeraddr=%s", pfrom.addr.ToStringAddrPort()) : "");
-            if (NUM_PRIVATE_BROADCAST_PER_TX > num_broadcasted.value()) {
-                // Not all of the initial NUM_PRIVATE_BROADCAST_PER_TX connections were needed.
-                // Tell CConnman it does not need to start the remaining ones.
-                m_connman.m_private_broadcast.NumToOpenSub(NUM_PRIVATE_BROADCAST_PER_TX - num_broadcasted.value());
-            }
-        }
-
         LOCK2(cs_main, m_tx_download_mutex);
 
         m_txrequest.ReceivedResponse(pfrom.GetId(), txid);
         if (tx.HasWitness()) m_txrequest.ReceivedResponse(pfrom.GetId(), wtxid);
 
         // We do the AlreadyHaveTx() check using wtxid, rather than txid - in the
@@ -4878,21 +4863,37 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             //
             // Note that m_lazy_recent_rejects doesn't just record DoSy or invalid
             // transactions, but any tx not accepted by the mempool, which may be
             // due to node policy (vs. consensus). So we can't blanket penalize a
             // peer simply for relaying a tx that our m_lazy_recent_rejects has caught,
             // regardless of false positives.
+            // XXX what about Remove() also here?
             return;
         }
 
         const MempoolAcceptResult result = m_chainman.ProcessTransaction(ptx);
         const TxValidationState& state = result.m_state;
 
         if (result.m_result_type == MempoolAcceptResult::ResultType::VALID) {
             ProcessValidTx(pfrom.GetId(), ptx, result.m_replaced_transactions);
             pfrom.m_last_tx_time = GetTime<std::chrono::seconds>();
+
+            if (auto num_broadcasted = m_tx_for_private_broadcast.Remove(ptx)) {
+                LogPrintLevel(BCLog::PRIVATE_BROADCAST,
+                              BCLog::Level::Info,
+                              "Received our privately broadcast transaction (txid=%s) from the "
+                              "network from peer=%d%s; stopping private broadcast attempts\n",
+                              txid.ToString(),
+                              pfrom.GetId(),
+                              fLogIPs ? strprintf(", peeraddr=%s", pfrom.addr.ToStringAddrPort()) : "");
+                if (NUM_PRIVATE_BROADCAST_PER_TX > num_broadcasted.value()) {
+                    // Not all of the initial NUM_PRIVATE_BROADCAST_PER_TX connections were needed.
+                    // Tell CConnman it does not need to start the remaining ones.
+                    m_connman.m_private_broadcast.NumToOpenSub(NUM_PRIVATE_BROADCAST_PER_TX - num_broadcasted.value());
+                }
+            }
         }
         else if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS)
         {
             bool fRejectedParents = false; // It may be the case that the orphans parents have all been rejected
 
             // Deduplicate parent txids, so that we don't have to loop over
```
</details>

Note that if the tx is not accepted in our mempool because it has too low fee for our mempool, but is accepted in everybody else's mempool, then we would do the remaining broadcasts up to `NUM_PRIVATE_BROADCAST_PER_TX` (5), then it would get stale and will be removed by `PeerManagerImpl::ReattemptPrivateBroadcast()`. Whereas currently, without this change it will be removed right away and we will not do the remaining broadcasts (up to 5).

I got another idea - remove the transaction only if the received tx it has the same txid and the same wtxid as the one that we broadcast.

<details>
<summary>[patch] check also wtxid</summary>

```diff
--- i/src/private_broadcast.cpp
+++ w/src/private_broadcast.cpp
@@ -21,13 +21,13 @@ bool PrivateBroadcast::Add(const CTransactionRef& tx) EXCLUSIVE_LOCKS_REQUIRED(!
 }
 
 std::optional<size_t> PrivateBroadcast::Remove(const CTransactionRef& tx) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
 {                 
     LOCK(m_mutex);
     auto iters = Find(tx->GetHash());
-    if (!iters) {
+    if (!iters || iters->by_txid->second.tx->GetWitnessHash() != tx->GetWitnessHash()) {
         return std::nullopt;
     }
     const size_t num_broadcasted{iters->by_priority->first.num_broadcasted};
     m_by_priority.erase(iters->by_priority);
     m_by_txid.erase(iters->by_txid);
     return num_broadcasted;
```
</details>

With this, if a malleated and valid transaction makes it to our mempool, then we would keep broadcasting the original one, up to 5 times, and then it will be removed by `PeerManagerImpl::ReattemptPrivateBroadcast()`.

Which one?

Reply:
> Which one?

No strong opinion (I think that both should work), but the second one (also check wtxid) seems simpler.

Reply:
I went for the second one being simpler and also easier to reason about - if we received something other than what we broadcasted, then keep broadcasting (but not infinitely).